url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/UART.java

```diff
@@ -0,0 +1,346 @@
+        return slave;
+    }
+
+    class UARTSlave extends Controller32 {
+        public UARTSlave() {
+            addReg(REG_UARTDR, "UARTDR", 0x00000000);
+            addReg(REG_UARTFR, "UARTFR", 0x00000000);
+
+            addReg(REG_UARTIBRD, "UARTIBRD", 0x00000000);
+            addReg(REG_UARTFBRD, "UARTFBRD", 0x00000000);
+            addReg(REG_UARTLCR_H, "UARTLCR_H", 0x00000000);
+            addReg(REG_UARTCR, "UARTCR", 0x00000000);
+            addReg(REG_UARTIFLS, "UARTIFLS", 0x00000000);
+
+            addReg(REG_UARTIMSC, "UARTIMSC", 0x00000000);
+            addReg(REG_UARTRIS, "UARTRIS", 0x00000000);
+            addReg(REG_UARTMIS, "UARTMIS", 0x00000000);
+            addReg(REG_UARTICR, "UARTICR", 0x00000000);
+
+            addReg(REG_UARTPeriphID0, "UARTPeriphID0", 0x00000011);
+            addReg(REG_UARTPeriphID1, "UARTPeriphID1", 0x00000010);
+            addReg(REG_UARTPeriphID2, "UARTPeriphID2", 0x00000014);
+            addReg(REG_UARTPeriphID3, "UARTPeriphID3", 0x00000000);
+            addReg(REG_UARTPCellID0, "UARTPCellID0", 0x0000000d);
+            addReg(REG_UARTPCellID1, "UARTPCellID1", 0x000000f0);
+            addReg(REG_UARTPCellID2, "UARTPCellID2", 0x00000005);
+            addReg(REG_UARTPCellID3, "UARTPCellID3", 0x000000b1);
+        }
+
+        @Override
+        public int readWord(long addr) {
+            int regaddr;
+            int result;
+
+            regaddr = (int)(addr & getAddressMask(LEN_WORD_BITS));
+
+            switch (regaddr) {
+            case REG_UARTDR:
+                if (bufInput.length() > 0) {
+                    result = bufInput.charAt(0);
+                    bufInput.deleteCharAt(0);
+                } else {
+                    result = 0;
+                }
+                break;
+            case REG_UARTFR:
+                result = 0;
+
+                //送信 FIFO は常に空いていることにする
+                result = BitOp.setBit32(result, FR_TXFE, true);
+                //受信 FIFO はバッファ残量に応じて設定する
+                result = BitOp.setBit32(result, FR_RXFE, bufInput.length() == 0);
+
+                break;
+            case REG_UARTLCR_H:
+                result = super.readWord(regaddr);
+                //System.out.printf("UARTLCR_H: read 0x%08x\n", result);
+                break;
+            case REG_UARTCR:
+                result = super.readWord(regaddr);
+                //System.out.printf("UARTCR: read 0x%08x\n", result);
+                break;
+            case REG_UARTIMSC:
+                result = maskInt;
+                break;
+            case REG_UARTRIS:
+                result = getRawInt();
+                break;
+            case REG_UARTMIS:
+                result = getMaskedInt();
+                break;
+            default:
+                result = super.readWord(regaddr);
+                break;
+            }
+
+            return result;
+        }
+
+        @Override
+        public void writeWord(long addr, int data) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 46 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 98% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.